variation-commons-mongodb/src/main/java/uk/ac/ebi/eva/commons/mongodb/writers/VariantMongoWriter.java

```diff
@@ -103,21 +105,22 @@ private void createIndexes() {
     protected void doWrite(List<? extends IVariant> variants) {
         BulkWriteOperation bulk = mongoOperations.getCollection(collection).initializeUnorderedBulkOperation();
         for (IVariant variant : variants) {
-            String id = VariantMongo.buildVariantId(variant.getChromosome(), variant.getStart(),
-                                                    variant.getReference(), variant.getAlternate());
-
-            // the chromosome and start appear just as shard keys, in an unsharded cluster they wouldn't be needed
-            BasicDBObject query = new BasicDBObject("_id", id)
-                    .append(VariantMongo.CHROMOSOME_FIELD, variant.getChromosome())
-                    .append(VariantMongo.START_FIELD, variant.getStart());
-
-            bulk.find(query).upsert().updateOne(generateUpdate(variant));
-
+            bulk.find(generateQuery(variant)).upsert().updateOne(generateUpdate(variant));
         }
 
         executeBulk(bulk, variants.size());
     }
 
+    private BasicDBObject generateQuery(IVariant variant) {
+        String id = VariantMongo.buildVariantId(variant.getChromosome(), variant.getStart(),
+                                                variant.getReference(), variant.getAlternate());
+
+        // the chromosome and start appear just as shard keys, in an unsharded cluster they wouldn't be needed
+        return new BasicDBObject("_id", id)
+                .append(VariantMongo.CHROMOSOME_FIELD, variant.getChromosome())
+                .append(VariantMongo.START_FIELD, variant.getStart());
+    }
+
     private void executeBulk(BulkWriteOperation bulk, int currentBulkSize) {
```

Comment:
According to the execution flow, this method should be moved after `generateUpdate` and related private methods.

Reply:
I was not sure the readability would increase if the executeBulk is moved so far away from its usage.